yt/frontends/nifti/io.py

```diff
@@ -0,0 +1,44 @@
+import numpy as np
+
+from yt.utilities.io_handler import BaseIOHandler
+from yt.utilities.on_demand_imports import _nibabel as nib
+
+
+class NiftiIOHandler(BaseIOHandler):
+    _particle_reader = False
+    _dataset_type = "nifti"
+
+    def _read_particle_coords(self, chunks, ptf):
+        # This needs to *yield* a series of tuples of (ptype, (x, y, z)).
+        # chunks is a list of chunks, and ptf is a dict where the keys are
+        # ptypes and the values are lists of fields.
+        raise NotImplementedError
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # This gets called after the arrays have been allocated.  It needs to
+        # yield ((ptype, field), data) where data is the masked results of
+        # reading ptype, field and applying the selector to the data read in.
+        # Selector objects have a .select_points(x,y,z) that returns a mask, so
+        # you need to do your masking here.
+        raise NotImplementedError
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
+
+        rv = {field: np.empty(size, dtype="float64") for field in fields}
+
+        offset = 0
+
+        for field in fields:
+            for chunk in chunks:
+                for grid in chunk.objs:
+                    variable = nib.load(self.ds.filename)
+                    data = variable.get_fdata()  # .astype("f8")
+                    offset += grid.select(selector, data, rv[field], offset)
```

Comment:
```suggestion
        variable = nib.load(self.ds.filename)
        data = variable.get_fdata()
        for field in fields:
            for chunk in chunks:
                for grid in chunk.objs:
                    offset += grid.select(selector, data, rv[field], offset)
```

Some of the operations performed within the loop are independent of loop iteration. Specifically opening the nifti file and reading all data from it, could be costly and should happen only once. There's non-zero chance that `nibabel` caches those things already, but let's err on the side of caution.